proxy/core/src/main/scala/io/cloudstate/proxy/crud/CrudEntity.scala

```diff
@@ -0,0 +1,415 @@
+
+  override final def receive: Receive = PartialFunction.empty
+
+  override final def preStart(): Unit = {
+    client
+      .handle(
+        Source
+          .actorRef[CrudStreamIn](configuration.sendQueueSize, OverflowStrategy.fail)
+          .mapMaterializedValue { ref =>
+            self ! Relay(ref)
+            NotUsed
+          }
+      )
+      .runWith(Sink.actorRef(self, CrudEntity.StreamClosed, CrudEntity.StreamFailed))
+    context.become(waitingForRelay)
+  }
+
+  private[this] final def waitingForRelay: Receive = {
+    case Relay(relayRef) =>
+      // Cluster sharding URL encodes entity ids, so to extract it we need to decode.
+      val entityId = URLDecoder.decode(self.path.name, "utf-8")
+      val manager = context.watch(
+        context
+          .actorOf(CrudEntity.props(configuration, entityId, relayRef, concurrencyEnforcer, statsCollector), "entity")
+      )
+      context.become(forwarding(manager, relayRef))
+      unstashAll()
+    case _ => stash()
+  }
+
+  private[this] final def forwarding(manager: ActorRef, relay: ActorRef): Receive = {
+    case Terminated(`manager`) =>
+      if (streamTerminated) {
+        context.stop(self)
+      } else {
+        relay ! Status.Success(CompletionStrategy.draining)
+        context.become(stopping)
+      }
+
+    case toParent if sender() == manager =>
+      context.parent ! toParent
+
+    case CrudEntity.StreamClosed =>
+      streamTerminated = true
+      manager forward CrudEntity.StreamClosed
+
+    case failed: CrudEntity.StreamFailed =>
+      streamTerminated = true
+      manager forward failed
+
+    case msg =>
+      manager forward msg
+  }
+
+  private def stopping: Receive = {
+    case CrudEntity.StreamClosed =>
+      context.stop(self)
+    case _: CrudEntity.StreamFailed =>
+      context.stop(self)
+  }
+
+  override def supervisorStrategy: SupervisorStrategy = SupervisorStrategy.stoppingStrategy
+}
+
+object CrudEntity {
+
+  final case object Stop
+
+  final case object StreamClosed extends DeadLetterSuppression
+  final case class StreamFailed(cause: Throwable) extends DeadLetterSuppression
+
+  final case class Configuration(
+      serviceName: String,
+      userFunctionName: String,
+      passivationTimeout: Timeout,
+      sendQueueSize: Int
+  )
+
+  private final case class OutstandingCommand(
+      commandId: Long,
+      actionId: String,
+      replyTo: ActorRef
+  )
+
+  final def props(configuration: Configuration,
+                  entityId: String,
+                  relay: ActorRef,
+                  concurrencyEnforcer: ActorRef,
+                  statsCollector: ActorRef): Props =
+    Props(new CrudEntity(configuration, entityId, relay, concurrencyEnforcer, statsCollector))
+
+  /**
+   * Used to ensure the action ids sent to the concurrency enforcer are indeed unique.
+   */
+  private val actorCounter = new AtomicLong(0)
+
+}
+
+final class CrudEntity(configuration: CrudEntity.Configuration,
```

Comment:
@ralphlaude @pvlugter @jroper I think the big thing left here is to implement native support for CRUD—i.e. not using Akka Persistence but rather direct storage CRUD support.

Reply:
@viktorklang it is good point with native support for CRUD. i would like to understand why native support for CRUD is a better fit than using Akka Persistence.

Reply:
@ralphlaude The thinking is that eventsourcing and snapshots might not be the optimal representation for a database geared towards CRUD.

Reply:
@viktorklang @ralphlaude Probably, but not only, this could be implemented on top of KV support
Of course, KV support itself must be implemented on top of some persistent storage (Postgres, Hazelcast, Infinispan, Redis, Riak, etc ...)

Reply:
@sleipnir Indeed! I've been toying with ideas on how to make storage engine per state model pluggable, but it might be premature at this point :)

Reply:
Cool. \0/

Reply:
@viktorklang, @sleipnir i agree! thanks, the idea for native support for CRUD makes sense now (we discussed it yesterday in the contributor call) :). it will be nice start supporting what we already have (JDBC, Postgres and Cassandra) :).
i will probably need some support here.